p4-16/spec/P4-16-spec.mdk

```diff
@@ -6137,8 +6167,8 @@ header, header stack, `struct`, or header union to the output packet.
 - When applied to a `struct` or header union, `emit` recursively
   invokes itself to each field.
 
-It is illegal to invoke `emit` on an expression of whose type is a
-base type, `enum`, or `error`.
+It is illegal to invoke `emit` on an expression whose type is
+`enum`, `error`, or a base type other than an integer (`int<W>`) or Boolean type.
```

Comment:
Is the intent here to change the spec so that `emit` is now explicitly _allowed_ if the expression has type `bit<W>`, `int<W>`, a serializable enum, boolean?  That is, emit directly on a value of those types, even if they are not inside of a header, or a struct inside of a header?

If so, I think it should be stated more clearly, and I can help with wording suggestions.  But first I wanted to check whether that was the intent.

Reply:
Good point.

The current spec does not allow any base type, enum, or error as an argument to the emit() method.   However, since an ```int<W>```  and ```bool``` can be serialized, one could allow them as arguments to the emit method.  But what would the use case be? 

For the other point you raised, indeed, it makes sense to clarify that when a struct is an argument to emit, the struct can only include certain types which are serializable fields.

Reply:
Before your changes, the spec explicitly said that emit was allowed on headers.  Your changes could be viewed as generalizing what can be _inside_ of a header, while still allowing emit (and extract) to work on them.

If so, and if there are no immediately plans to generalize emit to work on new types, is there any reason to change this sentence at all?

Reply:
Before the nested struct code changes, a header only allowed bit-fields and fixed-width signed integers.  With the changes, a nested struct is allowed, but the struct can include a kitchen sink of types some of which are enum and error which a header cannot support.   Thus, I thought a clarification would help.  However, the clarification may not be in the illegal sentence of section 15.1 - we could make the clarification elsewhere.  Or we could skip any clarifications altogether because it may be obvious that an error and enum are not serializable.

Reply:
Maybe I am reading it incorrectly, but when it says that it is illegal to invoke emit on an expression whose type is X, Y, or Z, I read that as the type of the entire expression cannot be X, Y, or Z.

It does not specify anything about whether the expression might contain inside of it, as part of its definition, a field with type X, Y, or Z.

Reply:
An alternate suggested change in the text, which is intended to mean exactly what it does now:

"It is illegal to invoke `emit` on an expression whose type is not one of those listed above."

The list above already explicitly says you can do `emit` on a header.  The changes you are making earlier in the spec explicitly say that a header can contain `bool`, `int<W>`, etc.